usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -489,6 +519,467 @@ static CK_RV cleanse_attribute(TEMPLATE *template,
+        0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10
+    };
+    CK_MECHANISM mech = { CKM_IBM_CPACF_WRAP, &iv, sizeof(iv) };
+    pkey_wrap_handler_data_t *data = (pkey_wrap_handler_data_t *) handler_data;
+    target_t target = 0;
+    CK_RV ret;
+
+    if (data->wrap_was_successful)
+        goto done;
+
+    ret = get_ep11_target_for_apqn(adapter, domain, &target, XCP_MFL_PROBE);
+    if (ret != CKR_OK)
+        goto done;
+
+    /* Create the protected key via CKM_IBM_CPACF_WRAP */
+    ret = dll_m_WrapKey(data->secure_key, data->secure_key_len,
+                        NULL, 0, NULL, 0, &mech,
+                        data->pkey_buf, data->pkey_buflen_p,
+                        target | XCP_TGTFL_SET_SCMD);
+    if (ret == CKR_OK)
+        data->wrap_was_successful = CK_TRUE;
+
+    free_ep11_target_for_apqn(target);
+
+done:
+
+    /* Always return ok, calling function loops over this handler until
+     * data->wrap_was_successful = true, or no more APQN left */
+    return CKR_OK;
+}
+
+/**
+ * Creates a protected key from the given secure key object via the ep11 lib
+ * CKM_IBM_CPACF_WRAP mechanism.
+ */
+static CK_RV ep11tok_pkey_skey2pkey(STDLL_TokData_t *tokdata,
+                                    CK_ATTRIBUTE *skey_attr,
+                                    CK_ATTRIBUTE **pkey_attr)
+{
+    ep11_private_data_t *ep11_data = tokdata->private_data;
+    CK_ATTRIBUTE *tmp_attr = NULL;
+    CK_BYTE ep11_buf[sizeof(wrapped_key_t)];
+    size_t ep11_buflen = sizeof(ep11_buf);
+    pkey_wrap_handler_data_t pkey_wrap_handler_data;
+    wrapped_key_t *wk;
+    CK_RV ret;
+
+    /* Create the protected key via CKM_IBM_CPACF_WRAP */
+    memset(&pkey_wrap_handler_data, 0, sizeof(pkey_wrap_handler_data_t));
+    pkey_wrap_handler_data.secure_key = skey_attr->pValue;
+    pkey_wrap_handler_data.secure_key_len = skey_attr->ulValueLen;
+    pkey_wrap_handler_data.pkey_buf = (CK_BYTE *)&ep11_buf;
+    pkey_wrap_handler_data.pkey_buflen_p = &ep11_buflen;
+    ret = handle_all_ep11_cards(&ep11_data->target_list, ep11tok_pkey_wrap_handler,
+                                &pkey_wrap_handler_data);
+    if (ret != CKR_OK || !pkey_wrap_handler_data.wrap_was_successful) {
+        TRACE_ERROR("handle_all_ep11_cards failed or no APQN could do the wrap.\n");
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Check ep11 wrapped key struct length */
+    if (ep11_buflen != sizeof(wrapped_key_t)) {
+        TRACE_ERROR("invalid ep11 wrapped key struct length %ld\n", ep11_buflen);
+        ret = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    /* Check if returned key type is what we expect:
+     *  - 0x1 = AES with bit length 128, 192 and 256
+     *  - 0x2 = 2DES and 3DES
+     *  - 0x3 = EC-P with bit length 192, 224, 256, 384 and 521
+     *  - 0x4 = ED25519 and ED448
+     */
+    wk = (wrapped_key_t *) &ep11_buf;
+    switch (wk->cpacf_key_type) {
+    case 1:
+    case 3:
```

Comment:
Are there symbolic names for these CPACF key types that we could use instead of the hard-coded constants?

Reply:
No, at least not yet. These "key type" values are part of the specification of the CKM_IBM_CPACF_WRAP mechanism, as provided by zFirmware. They provide a means to check if we get what we wanted. So they are not really "key types", like e.g. AES, but more like "indications" of certain groups of keys. Sure, I can introduce constants for them, but as they are only used here, I thought it's sufficient to explain their meaning in the comment.

Reply:
Yes, please introduce defines, to have readable code. The names of the defines should tell the reader what this value means.

Reply:
Done.